sel/src/Sel/PublicKey/AuthenticatedEncryption.hs

```diff
@@ -0,0 +1,494 @@
+    BS.fromForeignPtr0
+      (Foreign.castForeignPtr @CUChar @Word8 secretKeyForeignPtr)
+      (fromIntegral @CSize @Int cryptoBoxSecretKeyBytes)
+
+--
+
+-- | Convert a 'SecretKey' to a hexadecimal-encoded 'StrictByteString'.
+--
+-- ⚠️  Be prudent as to where you store it!
+--
+-- @since 0.0.1.0
+unsafePublicKeyToHexByteString :: PublicKey -> StrictByteString
+unsafePublicKeyToHexByteString (PublicKey publicKeyForeignPtr) =
+  Base16.encodeBase16' $
+    BS.fromForeignPtr0
+      (Foreign.castForeignPtr @CUChar @Word8 publicKeyForeignPtr)
+      (fromIntegral @CSize @Int cryptoBoxPublicKeyBytes)
+
+-- | A random number that must only be used once per exchanged message.
+-- It does not have to be confidential.
+-- It is of size 'cryptoBoxNonceBytes'.
+--
+-- @since 0.0.1.0
+newtype Nonce = Nonce (ForeignPtr CUChar)
+  deriving
+    ( Display
+      -- ^ @since 0.0.1.0
+      -- > display secretKey == "[REDACTED]"
+    )
+    via (ShowInstance Nonce)
+
+-- |
+--
+-- @since 0.0.1.0
+instance Eq Nonce where
+  (Nonce hk1) == (Nonce hk2) =
+    unsafeDupablePerformIO $
+      foreignPtrEq hk1 hk2 cryptoBoxNonceBytes
+
+-- |
+--
+-- @since 0.0.1.0
+instance Ord Nonce where
+  compare (Nonce hk1) (Nonce hk2) =
+    unsafeDupablePerformIO $
+      foreignPtrOrd hk1 hk2 cryptoBoxNonceBytes
+
+-- |
+--
+-- @since 0.0.1.0
+instance Show Nonce where
+  show = show . nonceToHexByteString
+
+-- | Generate a new random nonce.
+-- Only use it once per exchanged message.
+--
+-- Do not use this outside of hash creation!
+newNonce :: IO Nonce
+newNonce = do
+  (fPtr :: ForeignPtr CUChar) <- Foreign.mallocForeignPtrBytes (fromIntegral cryptoBoxNonceBytes)
+  Foreign.withForeignPtr fPtr $ \ptr ->
+    randombytesBuf (Foreign.castPtr @CUChar @Word8 ptr) cryptoBoxNonceBytes
+  pure $ Nonce fPtr
+
+-- | Create a 'Nonce' from a binary 'StrictByteString' that you have obtained on your own,
+-- usually from the network or disk.
+--
+-- @since 0.0.1.0
+nonceFromByteString :: StrictByteString -> Nonce
+nonceFromByteString bytestring = unsafeDupablePerformIO $
+  BS.unsafeUseAsCStringLen bytestring $ \(outsideNoncePtr, _) -> do
+    nonceForeignPtr <-
+      BS.mallocByteString
+        @CChar
+        (fromIntegral cryptoBoxNonceBytes)
+    Foreign.withForeignPtr nonceForeignPtr $ \noncePtr ->
+      Foreign.copyArray
+        outsideNoncePtr
+        noncePtr
+        (fromIntegral cryptoBoxNonceBytes)
+    pure $ Nonce (Foreign.castForeignPtr @CChar @CUChar nonceForeignPtr)
+
+-- | Convert a 'Nonce' to a hexadecimal-encoded 'StrictByteString'.
+--
+-- @since 0.0.1.0
+nonceToHexByteString :: Nonce -> StrictByteString
+nonceToHexByteString (Nonce nonceForeignPtr) =
+  Base16.encodeBase16' $
+    BS.fromForeignPtr0
+      (Foreign.castForeignPtr @CUChar @Word8 nonceForeignPtr)
+      (fromIntegral @CSize @Int cryptoBoxNonceBytes)
+
+-- | A ciphertext consisting of an encrypted message and an authentication tag.
+--
+-- @since 0.0.1.0
+data Hash = Hash
```

Comment:
It feels kind of strange to me to call a ciphertext a hash.

Apart from that, I think it might be a good idea to include the nonce with the message so that the association between the nonce and the message is not easily lost?

Reply:
Yep, renaming it.